File: lib/public-organisations.ts
export type OrganisationKind =
  | "PUBLIC_BODY"
  | "COMPANY"
  | "NON_PROFIT"
  | "EUROPEAN_BODY"
  | "OTHER";

export type PublicOrganisationSummary = {
  id: string;
  legal_name: string;
  kind: OrganisationKind;
  registry_record_id: string;
  public_record_sha256: string;
  published_at: string;
};

export type PublicOrganisationReply = {
  public_reference: string;
  original_record_sha256: string;
  claimant_public_name: string;
  claimant_role: string;
  statement_text: string;
  statement_sha256: string;
  official_response_url: string | null;
  submitted_at: string;
};

export type PublicOrganisation = PublicOrganisationSummary & {
  official_url: string;
  source_record_sha256: string;
  source: {
    publisher: "IRN";
    title: string;
    url: string;
    retrieved_at: string;
    content_sha256: string;
  };
  replies: PublicOrganisationReply[];
  coverage_notice: string;
};

export type PublicOrganisationHistory = {
  public_id: string;
  action: "PUBLISH" | "WITHDRAW";
  public_record_sha256: string;
  source_record_sha256: string;
  public_rationale: string;
  actor_alias: string;
  occurred_at: string;
};

export type PublicOrganisationHistoryResult = {
  public_id: string;
  items: PublicOrganisationHistory[];
  replies: PublicOrganisationReply[];
  coverage_notice: string;
};

type LoadResult<T> =
  | { status: "ok"; data: T }
  | { status: "not_found" | "unavailable" };

const idPattern = /^base_public_org_[0-9a-f]{32}$/;
const hashPattern = /^[0-9a-f]{64}$/;
const kinds = new Set<OrganisationKind>([
  "PUBLIC_BODY",
  "COMPANY",
  "NON_PROFIT",
  "EUROPEAN_BODY",
  "OTHER",
]);
export const ORGANISATION_KIND_LABELS: Record<OrganisationKind, string> = {
  PUBLIC_BODY: "Entidade pública",
  COMPANY: "Empresa",
  NON_PROFIT: "Organização sem fins lucrativos",
  EUROPEAN_BODY: "Entidade europeia",
  OTHER: "Outra organização",
};
const apiBase = process.env.NEXT_PUBLIC_API_URL?.replace(/\/$/, "");

function object(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function text(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value : null;
}

function timestamp(value: unknown): string | null {
  const result = text(value);
  return result !== null && Number.isFinite(Date.parse(result)) ? result : null;
}

function publicUrl(value: unknown, exactIrn = false): string | null {
  const result = text(value);
  if (result === null) return null;
  try {
    const url = new URL(result);
    if (
      url.protocol !== "https:" ||
      url.username ||
      url.password ||
      (exactIrn &&
        (url.hostname !== "publicacoes.mj.pt" ||
          url.port !== "" ||
          url.pathname !== "/DetalhePublicacao.aspx" ||
          url.search ||
          url.hash))
    ) {
      return null;
    }
    return result;
  } catch {
    return null;
  }
}

function summary(value: unknown): PublicOrganisationSummary | null {
  const item = object(value);
  if (item === null) return null;
  const id = text(item.id);
  const legalName = text(item.legal_name);
  const kind = text(item.kind);
  const registryRecordId = text(item.registry_record_id);
  const publicRecordSha256 = text(item.public_record_sha256);
  const publishedAt = timestamp(item.published_at);
  if (
    id === null ||
    !idPattern.test(id) ||
    legalName === null ||
    kind === null ||
    !kinds.has(kind as OrganisationKind) ||
    registryRecordId === null ||
    publicRecordSha256 === null ||
    !hashPattern.test(publicRecordSha256) ||
    publishedAt === null
  ) {
    return null;
  }
  return {
    id,
    legal_name: legalName,
    kind: kind as OrganisationKind,
    registry_record_id: registryRecordId,
    public_record_sha256: publicRecordSha256,
    published_at: publishedAt,
  };
}

function reply(value: unknown): PublicOrganisationReply | null {
  const item = object(value);
  if (item === null) return null;
  const publicReference = text(item.public_reference);
  const originalRecordSha256 = text(item.original_record_sha256);
  const claimantPublicName = text(item.claimant_public_name);
  const claimantRole = text(item.claimant_role);
  const statementText = text(item.statement_text);
  const statementSha256 = text(item.statement_sha256);
  const submittedAt = timestamp(item.submitted_at);
  const officialResponseUrl =
    item.official_response_url === null
      ? null
      : publicUrl(item.official_response_url);
  if (
    publicReference === null ||
    originalRecordSha256 === null ||
    !hashPattern.test(originalRecordSha256) ||
    claimantPublicName === null ||
    claimantRole === null ||
    statementText === null ||
    statementSha256 === null ||
    !hashPattern.test(statementSha256) ||
    submittedAt === null ||
    (item.official_response_url !== null && officialResponseUrl === null)
  ) {
    return null;
  }
  return {
    public_reference: publicReference,
    original_record_sha256: originalRecordSha256,
    claimant_public_name: claimantPublicName,
    claimant_role: claimantRole,
    statement_text: statementText,
    statement_sha256: statementSha256,
    official_response_url: officialResponseUrl,
    submitted_at: submittedAt,
  };
}

function historyItem(value: unknown): PublicOrganisationHistory | null {
  const item = object(value);
  if (item === null) return null;
  const publicId = text(item.public_id);
  const action = text(item.action);
  const publicRecordSha256 = text(item.public_record_sha256);
  const sourceRecordSha256 = text(item.source_record_sha256);
  const publicRationale = text(item.public_rationale);
  const actorAlias = text(item.actor_alias);
  const occurredAt = timestamp(item.occurred_at);
  if (
    publicId === null ||
    !idPattern.test(publicId) ||
    (action !== "PUBLISH" && action !== "WITHDRAW") ||
    publicRecordSha256 === null ||
    !hashPattern.test(publicRecordSha256) ||
    sourceRecordSha256 === null ||
    !hashPattern.test(sourceRecordSha256) ||
    publicRationale === null ||
    actorAlias === null ||
    occurredAt === null
  ) {
    return null;
  }
  return {
    public_id: publicId,
    action,
    public_record_sha256: publicRecordSha256,
    source_record_sha256: sourceRecordSha256,
    public_rationale: publicRationale,
    actor_alias: actorAlias,
    occurred_at: occurredAt,
  };
}

async function request(
  path: string,
): Promise<{ status: number; body: unknown } | null> {
  if (!apiBase) return null;
  try {
    const response = await fetch(`${apiBase}${path}`, {
      headers: { Accept: "application/json" },
      cache: "no-store",
      signal: AbortSignal.timeout(8_000),
    });
    return {
      status: response.status,
      body: await response.json().catch(() => null),
    };
  } catch {
    return null;
  }
}

export async function loadPublicOrganisations(
  page: number,
  limit = 24,
): Promise<
  LoadResult<{
    items: PublicOrganisationSummary[];
    total: number;
    limit: number;
    offset: number;
    coverage_notice: string;
  }>
> {
  const safeLimit =
    Number.isSafeInteger(limit) && limit >= 1 && limit <= 100 ? limit : 24;
  const maximumPage = Math.floor(10_000 / safeLimit) + 1;
  const safePage =
    Number.isSafeInteger(page) && page > 0
      ? Math.min(page, maximumPage)
      : 1;
  const response = await request(
    `/api/v1/public/organisations?limit=${safeLimit}&offset=${
      (safePage - 1) * safeLimit
    }`,
  );
  const body = object(response?.body);
  if (response?.status !== 200 || body === null || !Array.isArray(body.items)) {
    return { status: "unavailable" };
  }
  const items = body.items.map(summary);
  const total = body.total;
  const returnedLimit = body.limit;
  const offset = body.offset;
  const coverageNotice = text(body.coverage_notice);
  if (
    items.some((item) => item === null) ||
    !Number.isSafeInteger(total) ||
    Number(total) < 0 ||
    returnedLimit !== safeLimit ||
    offset !== (safePage - 1) * safeLimit ||
    coverageNotice === null
  ) {
    return { status: "unavailable" };
  }
  return {
    status: "ok",
    data: {
      items: items as PublicOrganisationSummary[],
      total: Number(total),
      limit: safeLimit,
      offset: Number(offset),
      coverage_notice: coverageNotice,
    },
  };
}

export async function loadPublicOrganisation(
  publicId: string,
): Promise<LoadResult<PublicOrganisation>> {
  if (!idPattern.test(publicId)) return { status: "not_found" };
  const response = await request(
    `/api/v1/public/organisations/${encodeURIComponent(publicId)}`,
  );
  if (response === null) return { status: "unavailable" };
  if (response.status === 404) return { status: "not_found" };
  const body = object(response.body);
  const base = summary(body);
  const source = object(body?.source);
  const sourceTitle = text(source?.title);
  const sourceUrl = publicUrl(source?.url, true);
  const retrievedAt = timestamp(source?.retrieved_at);
  const sourceContentSha256 = text(source?.content_sha256);
  const officialUrl = publicUrl(body?.official_url, true);
  const sourceRecordSha256 = text(body?.source_record_sha256);
  const coverageNotice = text(body?.coverage_notice);
  const replies = Array.isArray(body?.replies)
    ? body.replies.map(reply)
    : null;
  if (
    response.status !== 200 ||
    body === null ||
    base === null ||
    source === null ||
    source.publisher !== "IRN" ||
    sourceTitle === null ||
    sourceUrl === null ||
    retrievedAt === null ||
    sourceContentSha256 === null ||
    !hashPattern.test(sourceContentSha256) ||
    officialUrl === null ||
    sourceRecordSha256 === null ||
    !hashPattern.test(sourceRecordSha256) ||
    coverageNotice === null ||
    replies === null ||
    replies.some((item) => item === null)
  ) {
    return { status: "unavailable" };
  }
  return {
    status: "ok",
    data: {
      ...base,
      official_url: officialUrl,
      source_record_sha256: sourceRecordSha256,
      source: {
        publisher: "IRN",
        title: sourceTitle,
        url: sourceUrl,
        retrieved_at: retrievedAt,
        content_sha256: sourceContentSha256,
      },
      replies: replies as PublicOrganisationReply[],
      coverage_notice: coverageNotice,
    },
  };
}

export async function loadPublicOrganisationHistory(
  publicId: string,
): Promise<LoadResult<PublicOrganisationHistoryResult>> {
  if (!idPattern.test(publicId)) return { status: "not_found" };
  const response = await request(
    `/api/v1/public/organisations/${encodeURIComponent(
      publicId,
    )}/publication-history`,
  );
  if (response === null) return { status: "unavailable" };
  if (response.status === 404) return { status: "not_found" };
  const body = object(response.body);
  if (
    response.status !== 200 ||
    body === null ||
    !Array.isArray(body.items) ||
    !Array.isArray(body.replies)
  ) {
    return { status: "unavailable" };
  }
  const returnedPublicId = text(body.public_id);
  const coverageNotice = text(body.coverage_notice);
  const items = body.items.map(historyItem);
  const replies = body.replies.map(reply);
  const snapshotHashes = new Set(
    items.flatMap((item) => (item ? [item.public_record_sha256] : [])),
  );
  if (
    returnedPublicId !== publicId ||
    coverageNotice === null ||
    items.some((item) => item === null || item.public_id !== publicId) ||
    replies.some(
      (item) =>
        item === null || !snapshotHashes.has(item.original_record_sha256),
    )
  ) {
    return { status: "unavailable" };
  }
  return {
    status: "ok",
    data: {
      public_id: publicId,
      items: items as PublicOrganisationHistory[],
      replies: replies as PublicOrganisationReply[],
      coverage_notice: coverageNotice,
    },
  };
}

File: backend/app/repositories/public_organisations.py
"""Leitura explícita das fotografias V5.53, sem qualquer recuo para dados legados."""

from __future__ import annotations

import re
from typing import Any

import asyncpg


class PublicOrganisationRepository:
    def __init__(self, pool: asyncpg.Pool | None) -> None:
        self.pool = pool

    def _require_pool(self) -> asyncpg.Pool:
        if self.pool is None:
            raise RuntimeError("Base de dados não configurada")
        return self.pool

    async def list(self, *, limit: int, offset: int) -> dict[str, Any]:
        pool = self._require_pool()
        where = """o.id ~ '^base_public_org_[0-9a-f]{32}$'
          AND o.publication_status = 'PUBLISHED' AND o.verification_status = 'VERIFIED'
          AND o.current_publication_snapshot_id = s.id
          AND EXISTS (SELECT 1 FROM editorial_publication_events e
            WHERE e.case_id=s.editorial_case_id AND e.version_id=s.editorial_version_id
              AND e.target_type='BASE_PUBLIC_ORGANISATION' AND e.target_id=o.id
              AND e.action='PUBLISH')"""
        async with (
            pool.acquire() as connection,
            connection.transaction(isolation="repeatable_read", readonly=True),
        ):
            total = await connection.fetchval(
                f"""SELECT COUNT(*)::int FROM organisations o
                     JOIN base_public_organisation_publication_snapshots s
                       ON s.organisation_id=o.id
                     WHERE {where}"""
            )
            rows = await connection.fetch(
                f"""SELECT o.id,s.legal_name,s.kind::text,s.registry_record_id,
                            s.public_record_sha256,s.created_at AS published_at
                     FROM organisations o
                     JOIN base_public_organisation_publication_snapshots s
                       ON s.organisation_id=o.id
                     WHERE {where}
                     ORDER BY s.legal_name COLLATE "C",o.id COLLATE "C"
                     LIMIT $1 OFFSET $2""",
                limit,
                offset,
            )
        return {
            "items": [dict(row) for row in rows],
            "total": int(total),
            "limit": limit,
            "offset": offset,
        }

    async def get(self, *, public_id: str) -> dict[str, Any] | None:
        if not re.fullmatch(r"base_public_org_[0-9a-f]{32}", public_id):
            return None
        pool = self._require_pool()
        async with (
            pool.acquire() as connection,
            connection.transaction(isolation="repeatable_read", readonly=True),
        ):
            row = await connection.fetchrow(
                """SELECT o.id,s.legal_name,s.kind::text,s.registry_record_id,s.official_url,
                      s.source_record_sha256,s.public_record_sha256,s.created_at AS published_at,
                      source.title AS source_title,source.url AS source_url,
                      source.retrieved_at,source.content_sha256
               FROM organisations o
               JOIN base_public_organisation_publication_snapshots s
                 ON s.id=o.current_publication_snapshot_id AND s.organisation_id=o.id
               JOIN source_documents source ON source.id=s.source_document_id
               WHERE o.id=$1 AND o.publication_status='PUBLISHED'
                 AND o.verification_status='VERIFIED' AND source.publisher='JUSTICE_REGISTRY'
                 AND EXISTS (SELECT 1 FROM editorial_publication_events e
                   WHERE e.case_id=s.editorial_case_id AND e.version_id=s.editorial_version_id
                     AND e.target_type='BASE_PUBLIC_ORGANISATION' AND e.target_id=o.id
                     AND e.action='PUBLISH')""",
                public_id,
            )
            if row is None:
                return None
            replies = await connection.fetch(
                """SELECT public_reference,original_record_sha256,
                          claimant_public_name,claimant_role,
                      statement_text,statement_sha256,official_response_url,submitted_at
               FROM rights_of_reply WHERE target_type='ORGANISATION' AND target_id=$1
                 AND original_record_sha256=$2 AND status='PUBLISHED'
               ORDER BY submitted_at,id""",
                public_id,
                row["public_record_sha256"],
            )
        return {
            "id": row["id"],
            "legal_name": row["legal_name"],
            "kind": row["kind"],
            "registry_record_id": row["registry_record_id"],
            "official_url": row["official_url"],
            "source_record_sha256": row["source_record_sha256"],
            "public_record_sha256": row["public_record_sha256"],
            "published_at": row["published_at"],
            "source": {
                "publisher": "IRN",
                "title": row["source_title"],
                "url": row["source_url"],
                "retrieved_at": row["retrieved_at"],
                "content_sha256": row["content_sha256"],
            },
            "replies": [dict(item) for item in replies],
        }

    async def history(self, *, public_id: str) -> dict[str, Any] | None:
        if not re.fullmatch(r"base_public_org_[0-9a-f]{32}", public_id):
            return None
        pool = self._require_pool()
        async with (
            pool.acquire() as connection,
            connection.transaction(isolation="repeatable_read", readonly=True),
        ):
            exists = await connection.fetchval(
                """SELECT EXISTS(
                       SELECT 1 FROM base_public_organisation_publication_snapshots
                       WHERE organisation_id=$1
                   )""",
                public_id,
            )
            if not exists:
                return None
            rows = await connection.fetch(
                """SELECT e.target_id AS public_id,e.action::text,s.public_record_sha256,
                      s.source_record_sha256,e.rationale AS public_rationale,
                      e.actor_alias,e.created_at AS occurred_at
               FROM editorial_publication_events e
               JOIN base_public_organisation_publication_snapshots s
                 ON s.editorial_case_id=e.case_id AND s.editorial_version_id=e.version_id
                AND s.organisation_id=e.target_id
               WHERE e.target_type='BASE_PUBLIC_ORGANISATION' AND e.target_id=$1
               ORDER BY e.created_at,e.id""",
                public_id,
            )
            replies = await connection.fetch(
                """SELECT reply.public_reference,reply.original_record_sha256,
                          reply.claimant_public_name,reply.claimant_role,
                          reply.statement_text,reply.statement_sha256,
                          reply.official_response_url,reply.submitted_at
                   FROM rights_of_reply reply
                   JOIN base_public_organisation_publication_snapshots snapshot
                     ON snapshot.organisation_id=reply.target_id
                    AND snapshot.public_record_sha256=reply.original_record_sha256
                   WHERE reply.target_type='ORGANISATION' AND reply.target_id=$1
                     AND reply.status='PUBLISHED'
                   ORDER BY reply.submitted_at,reply.id""",
                public_id,
            )
        return {
            "public_id": public_id,
            "items": [dict(row) for row in rows],
            "replies": [dict(row) for row in replies],
        }
